### PR TITLE
Low rank circuits

### PR DESCRIPTION
**Pull request by babbush**
Adding final function that will go in OpenFermion core to help assist with low rank circuit simulation.

### CODE REVIEW

**Comment by babbush** [comment]
@jarrodmcc I am looking into this now, but an issue missed by tests on my last PR is that the one-body squares output by the low_rank_two_body_decomposition function are not necessarily Hermitian. That makes it less clear how we should diagonalize things.

**Comment by babbush** [comment]
The issue seems to be fixed for an even number of qubits. More substantial tests will be forthcoming once these routines are used in the forthcoming circuit compilation routines.

File: src/openfermion/utils/_low_rank.py [hunk 1]
```diff
@@ -174,3 +174,35 @@ def low_rank_two_body_decomposition(chemist_two_body_coefficients,
...
+
+    Returns:
+        density_density_matrix(ndarray of floats) an N by N array storing
+            the diagonal two-body coefficeints :math:`V_{pq}` above.
+        basis_transformation_matrix (ndarray of floats) an N by N array
+            storing the values of the basis transformation.
+    """
+    # Diagonalize the one-body matrix.
+    eigenvalues, eigenvectors = numpy.linalg.eig(one_body_matrix)
+    basis_transformation_matrix = eigenvectors.transpose()
```
Comment (jarrodmcc) [inline src/openfermion/utils/_low_rank.py]: What's the convention being used that it's the transpose, but not original or conjugate transpose?  Seems like generally there could be complex eigenvectors with non-hermitian matrices

File: src/openfermion/utils/_low_rank_test.py [hunk 2]
```diff
@@ -190,3 +191,56 @@ def test_rank_reduction(self):
...
+
+    def test_one_body_square_decomposition(self):
+
+        # Initialize a random two-body FermionOperator.
+        n_qubits = 4
+        random_operator = get_fermion_operator(
+            random_interaction_operator(n_qubits))
+
+        # Convert to chemist tensor.
+        constant, one_body_coefficients, chemist_tensor = \
```
Comment (jarrodmcc) [inline src/openfermion/utils/_low_rank_test.py]: No backslashes to be consistent

File: src/openfermion/utils/_low_rank_test.py [hunk 3]
```diff
@@ -190,3 +191,56 @@ def test_rank_reduction(self):
...
+            self.assertAlmostEqual(0., numpy.amax(numpy.absolute(difference)))
+            density_density_alternative = numpy.outer(diagonal, diagonal)
+            difference = density_density_alternative - density_density_matrix
+            self.assertAlmostEqual(0., numpy.amax(numpy.absolute(difference)))
+
+            # Test spectra.
+            one_body_squared_spectrum = eigenspectrum(one_body_squared)
+            two_body_spectrum = eigenspectrum(two_body_operator)
+            difference = two_body_spectrum - one_body_squared_spectrum
+            self.assertAlmostEqual(0., numpy.amax(numpy.absolute(difference)))
```
Comment (jarrodmcc) [inline src/openfermion/utils/_low_rank_test.py]: If we expect the resulting normal ordered operators to be the same also, maybe we can test that as well? Or is that too onerous?

**Pull request merged**